cip/1.accepted/CIP2016-12-14-Constraint-syntax.adoc

```diff
@@ -0,0 +1,310 @@
+
+Alternatively, we could extend our previous constraint definition with this new requirement:
+
+[source, cypher]
+----
+CREATE CONSTRAINT color_schema
+FOR (c:Color)
+REQUIRE UNIQUE c.rgb, c.name
+REQUIRE exists(c.rgb)
+----
+
+This composite constraint will make sure that all `:Color` nodes has a value for their `rgb` property, and that its value is unique for each `name`.
+
+More complex constraint definitions are considered below:
+
+.Property value limitations
+[source, cypher]
+----
+CREATE CONSTRAINT road_width
+FOR ()-[r:ROAD]-()
+REQUIRE 5 < r.width < 50
+----
+
+.Cardinality
+[source, cypher]
+----
+CREATE CONSTRAINT spread_the_love
+FOR (p:Person)
+REQUIRE size((p)-[:LOVES]->()) > 3
+----
+
+.Endpoint requirements
+[source, cypher]
+----
+CREATE CONSTRAINT can_only_own_things
+FOR ()-[:OWNS]->(t)
+REQUIRE (t:Vehicle) OR (t:Building) OR (t:Object)
+----
+
+.Label coexistence
+[source, cypher]
+----
+CREATE CONSTRAINT programmers_are_people_too
+FOR (p:Programmer)
+REQUIRE p:Person
+----
+
+.Constraint example from CIR-2017-173
+[source, cypher]
+----
+CREATE CONSTRAINT enforce_dag_unique_for_R_links
+FOR p = (a)-[:R]-(b)
+REQUIRE length(p) < 2
+----
+
+Assuming a function `acyclic()` that takes a path as argument and returns `true` if and only if the same node does not appear twice in the path, otherwise `false`, we may express:
+
+.Constraint example from CIR-2017-172
+[source, cypher]
+----
+CREATE CONSTRAINT enforce_dag_acyclic_for_R_links
+FOR p = ()-[:R*]-()
+REQUIRE acyclic(p)
+----
+
+=== Interaction with existing features
+
+The main interaction between the constraints and the rest of the language occurs during updating statements.
+Existing constraints will cause any updating statements to fail, thereby fulfilling the main purpose of this feature.
+
+=== Alternatives
+
+Alternative syntaxes have been discussed:
+
+* `GIVEN`, `CONSTRAIN`, `ASSERT` instead of `FOR`
+* `ASSERT`, `ENFORCE`, `IMPLIES` instead of `REQUIRE`
+
+The use of an existing expression to express uniqueness -- instead of using the operator `UNIQUE` -- becomes unwieldy for multiple properties, as exemplified by the following:
+----
+FOR (p:Person), (q:Person)
+REQUIRE p.email <> q.email AND p <> q
+----
+
+== What others do
+
+In SQL, the following constraints exist (inspired by http://www.w3schools.com/sql/sql_constraints.asp):
+
+* `NOT NULL` - Indicates that a column cannot store a null value.
+* `UNIQUE` - Ensures that each row for a column must have a unique value.
+* `PRIMARY KEY` - A combination of a `NOT NULL` and `UNIQUE`. Ensures that a column (or a combination of two or more columns) has a unique identity, reducing the resources required to locate a specific record in a table.
+* `FOREIGN KEY` - Ensures the referential integrity of the data in one table matches values in another table.
+* `CHECK` - Ensures that the value in a column meets a specific condition
```

Comment:
Maybe we should have a leading keyword like that for non-UNIQUE constraints as well?